### DISCUSSION TITLE
Swayer manifesto

### DISCUSSION THREAD

**Discussion by serhii-freeman** [metarhia/swayer#25] [Ideas]
Предлагаю обсудить и в итоге создать манифест проекта по типу ["Манифест технологического стека Метархия"](https://github.com/metarhia/Manifesto/blob/master/RU/Technology.md).

Думаю будет полезно обсудить какую нишу может занять данный фреймворк и какую боль он может решать для каждого из сообщества. Так или иначе, каждый принимает решение об участии в проекте на основании картинки будущего. Давайте нарисуем эту картину.

**Comment by rohiievych** [comment]
Уже есть достаточно подробная документация по фреймворку, поэтому будет лучше, если все хорошо подумают, как её можно улучшить. Сейчас я не вижу смысла размножать файлы с описанием, так как разработчикам удобнее всего читать доку в одном месте, сейчас это [README](https://github.com/metarhia/swayer/blob/main/README.md), а в будущем это может быть оформлено под сайт. Можно даже на нём и написать этот сайт с документацией, как это делает [Angular](https://angular.io/).

Для чего нужен Swayer - разрабатывать быстрый, контролируемый и консистентый UI. Его особенность в том, что с его помощью можно создавать передаваемые наборы данных, которые будут разворачиваться в интерфейс. Проще говоря: скормил специальный JSON - получил браузерное веб приложение. В будущем, возможно, это будет хорошей базой для графического дизайнера сайтов, так как любой аспект приложения конфигурируется данными.

**Comment by serhii-freeman** [comment]
Окей. Тогда просто подискутируем тут.

Что мне нравится:
- Без зависимостей.
- Идея - JS везде. Именно идея очень нравится. Выглядит это непривычно и громоздко (DOM);
- JIT (не люблю сборщики, а кто их любит);
- Ожидаемая полная совместимость с metarhia/impress.
- Обещание "хорошей" архитектуры, а именно примеры с хорошей архитектурой.
- Lazy loading.
- Маленький размер.

Что мне хотелось бы:
- Возможность использовать без сборщиков.     |
- Встраивание на страницу в нужный элемент.   |  Короче нужен jQuery с архитектурой ))
- Возможность общения нескольких инстэнсов на одной странице. Не все есть SPA.
- Быстрая разработка в связке с impress. Наборы компонентов с валидаторами полей и т.п. Считаю что преимущества данного фрейма может раскрываться в связке с impress.
- Более императивный или очень простые декларативные абстракции. Кому захочется учить еще один фреймворк полный декларативности? Декларативность должна быть интуитивно понятной.
- Хороший архитектурный гайд или другие механизмы как не выстрелить себе в ногу.


Хоть мне нравится идея я не согласен что JS везде улучшит производительность труда.
Сравнение человека и компьютера в плане переключения контекста не корректно. Да и ресурсов тратиться очень мало на то самое переключение если оно происходит не часто и мало тредов)
Мозг работает иначе, ему офигенно когда разнообразие, это доказала мнемотехника.
Тут же, когда мы пишем стили, мы пишем css, конвертим его в camelCase, конвертим в json. Пишем html конвертим его в json. Мы не уходим от htm, css а конвертим их в json.
Мне такой подход нравится но это говорит мое иррациональное.
Для бизнеса html и css удобнее ведь можно посадить верстальщика который будет пилить компоненты не обучая его что такое json.
Короче честно будет будет сказать мы так делаем потомушо мы так хотим)

**Reply by rohiievych** [reply]
Что нравится:
  - Без зависимостей - _да, в стиле Метархии_;
  - Идея - JS везде. Именно идея очень нравится. Выглядит это непривычно и громоздко (DOM) - _это свобода для построения приложения на основе данных_;
  - JIT (не люблю сборщики, а кто их любит) - _для SSR всё-таки понадобится сделать компилятор + регидрацию, чтобы добиться максимально лёгкого приложения, например, для мобильных устройств, но это опционально_;
  - Ожидаемая полная совместимость с metarhia/impress - _будет_;
  - Обещание "хорошей" архитектуры, а именно примеры с хорошей архитектурой - _над этим нужно хорошо подумать, чтобы не переборщить_;
  - Lazy loading - _хорошая штука_;
  - Маленький размер - _приятно_;

Что хотелось бы:
  - Возможность использовать без сборщиков - _такая возможность сейчас есть, просто нужно подключить модуль_;
  - Встраивание на страницу в нужный элемент. | Короче нужен jQuery с архитектурой )) - _планирую сделать_;
  - Возможность общения нескольких инстэнсов на одной странице. Не все есть SPA - _фреймворк загрузился, закешировался и норм_;
  - Быстрая разработка в связке с impress. Наборы компонентов с валидаторами полей и т.п. Считаю что преимущества данного фрейма может раскрываться в связке с impress - _будет_;
  - Более императивный или очень простые декларативные абстракции. Кому захочется учить еще один фреймворк полный декларативности? Декларативность должна быть интуитивно понятной - _полная декларативность в виде только данных - это конечные абстракции над swayer, на низком уровне соблюдается баланс: есть императивные части, а есть декларативные_;
  - Хороший архитектурный гайд или другие механизмы как не выстрелить себе в ногу - _есть примеры, их можно улучшать, полноценный гайд тоже добавим в будущем_.
  
  
> не согласен что JS везде улучшит производительность труда

А где про это написано? )


По поводу единого синтаксиса - в документации это поясняется. Естественно, что все привыкли к XML-like синтаксису HTML, но отличие от JSON-like - визуальное. Что первое, что второе - это описание структуры документа, но json, в отличие от xml, совместим с js. В итоге в любом случае конечный код - HTML/CSS/JS, потому что так работает браузер и не иначе.
Верстальщик точно также может верстать и на json, но опять же, мы не гонимся угодить бизнесу.
И в общем то да, я просто захотел реализовать свою концепцию и за несколько месяцев написал этот фреймворк. Так что кому не нравится - тот просто проходит мимо )
Всё открыто для улучшений сообществом.

**Comment by serhii-freeman** [comment]
Я тоже за json - это красиво и упрощает JIT.
На этом поставим точку говоря словами классика - стадион так стадион - json так json)
По другим пунктам как будет время интересно мнение сообщества.
Напр. встраивание на страницу в нужный элемент как vue думаю важный момент для нового фреймворка. Что-то большое писать мало кто рискнет, а виджет на сайт это приемлемый вариант.
